File: dotnet/src/Envdoctor/Detectors/TypoDetector.cs
using Envdoctor.Models;
using static Envdoctor.Detectors.DetectorHelpers;

namespace Envdoctor.Detectors;

/// Typo detector: pairs names that are referenced but not defined with names
/// that are defined but not referenced, and have a small edit distance.
public sealed class TypoDetector : IDetector
{
    public string Id => "typo";
    public string Name => "typo";
    public string Description =>
        "A referenced variable name is very similar to a defined variable name and may be a typo.";

    private static int Levenshtein(string a, string b)
    {
        var matrix = new int[b.Length + 1, a.Length + 1];
        for (var i = 0; i <= b.Length; i++)
            matrix[i, 0] = i;
        for (var j = 0; j <= a.Length; j++)
            matrix[0, j] = j;

        for (var i = 1; i <= b.Length; i++)
        {
            for (var j = 1; j <= a.Length; j++)
            {
                var cost = b[i - 1] == a[j - 1] ? 0 : 1;
                matrix[i, j] = Math.Min(
                    Math.Min(matrix[i - 1, j] + 1, matrix[i, j - 1] + 1),
                    matrix[i - 1, j - 1] + cost);
            }
        }
        return matrix[b.Length, a.Length];
    }

    /// Earliest (file, line) a name is referenced at, across usages and
    /// compose/action definitions.
    private static (string Path, int Line) ReferenceSortKey(IndexedModel index, string name)
    {
        var origins = new List<Origin>();
        if (index.Usages.TryGetValue(name, out var u))
            origins.AddRange(u);
        if (index.ComposeDefinitions.TryGetValue(name, out var cd))
            origins.AddRange(cd.Select(d => d.Origin));
        if (index.ActionDefinitions.TryGetValue(name, out var ad))
            origins.AddRange(ad.Select(d => d.Origin));
        return OriginSortKey(origins);
    }

    private static bool IsLikelyTypo(string a, string b)
    {
        if (a == b)
            return false;
        if (a.Length < 4 || b.Length < 4)
            return false;
        var distance = Levenshtein(a, b);
        var minLen = Math.Min(a.Length, b.Length);
        // Distance of 1 is always flagged for names >= 4 chars.
        // Distance of 2 is flagged for names >= 6 chars.
        // Larger distances only when names are long and ratio is low.
        if (distance == 1)
            return true;
        if (distance == 2)
            return minLen >= 6;
        if (distance == 3)
            return minLen >= 10;
        return false;
    }

    public List<Finding> Detect(IndexedModel index)
    {
        var findings = new List<Finding>();

        var defined = index.EnvDefinitions.Keys.ToHashSet();
        var used = new HashSet<string>(index.Usages.Keys);
        used.UnionWith(index.ComposeDefinitions.Keys);
        used.UnionWith(index.ActionDefinitions.Keys);

        // Names referenced but not defined anywhere, ordered by where they are
        // first referenced (usages, then compose/action defs).
        var undefinedNames = used.Where(n => !defined.Contains(n)).ToList();
        undefinedNames.Sort((a, b) =>
        {
            var cmp = CompareKeys(ReferenceSortKey(index, a), ReferenceSortKey(index, b));
            return cmp != 0 ? cmp : string.CompareOrdinal(a, b);
        });
        // Names defined but never referenced anywhere, ordered by parse order.
        var unusedNames = index.EnvDefinitions.Keys.Where(n => !used.Contains(n)).ToList();
        unusedNames.Sort((a, b) =>
        {
            var ka = index.EnvDefinitions.TryGetValue(a, out var da) ? DefSortKey(da) : default;
            var kb = index.EnvDefinitions.TryGetValue(b, out var db) ? DefSortKey(db) : default;
            var cmp = CompareKeys(ka, kb);
            return cmp != 0 ? cmp : string.CompareOrdinal(a, b);
        });

        var seen = new HashSet<string>();

        foreach (var undefinedName in undefinedNames)
        {
            foreach (var unusedName in unusedNames)
            {
                if (!IsLikelyTypo(undefinedName, unusedName))
                    continue;
                var pairNames = new[] { undefinedName, unusedName };
                Array.Sort(pairNames, StringComparer.Ordinal);
                var pairKey = string.Join("\0", pairNames);
                if (seen.Add(pairKey))
                {
                    var origins = index.Usages.TryGetValue(undefinedName, out var o)
                        ? o.Select(x => x.Clone()).Take(3).ToList()
                        : new List<Origin>();
                    findings.Add(MakeFinding(
                        "typo",
                        Severity.Warning,
                        undefinedName,
                        $"did you mean \"{unusedName}\"? ({undefinedName} is referenced but not defined, {unusedName} is defined but unused)",
                        origins));
                }
            }
        }

        return findings;
    }
}

File: swift/Sources/EnvdoctorKit/Detectors/TypoDetector.swift
import Foundation

/// Typo detector: pairs names that are referenced but not defined with names
/// that are defined but not referenced, and have a small edit distance.
public struct TypoDetector: Detector {
    public let id = "typo"
    public let name = "typo"
    public let description = "A referenced variable name is very similar to a defined variable name and may be a typo."

    public init() {}

    private func levenshtein(_ a: String, _ b: String) -> Int {
        let a = Array(a)
        let b = Array(b)
        var matrix = [[Int]](repeating: [Int](repeating: 0, count: a.count + 1), count: b.count + 1)
        for i in 0...b.count { matrix[i][0] = i }
        for j in 0...a.count { matrix[0][j] = j }
        if a.isEmpty { return b.count }
        if b.isEmpty { return a.count }
        for i in 1...b.count {
            for j in 1...a.count {
                let cost = b[i - 1] == a[j - 1] ? 0 : 1
                matrix[i][j] = min(
                    matrix[i - 1][j] + 1,
                    matrix[i][j - 1] + 1,
                    matrix[i - 1][j - 1] + cost
                )
            }
        }
        return matrix[b.count][a.count]
    }

    private func isLikelyTypo(_ a: String, _ b: String) -> Bool {
        if a == b { return false }
        if a.count < 4 || b.count < 4 { return false }
        let distance = levenshtein(a, b)
        let minLen = min(a.count, b.count)
        if distance == 1 { return true }
        if distance == 2 { return minLen >= 6 }
        if distance == 3 { return minLen >= 10 }
        return false
    }

    public func detect(_ index: IndexedModel) -> [Finding] {
        let defined = Set(index.envDefinitions.map(\.0))
        // JS: new Set([...usages.keys(), ...composeDefinitions.keys(), ...actionDefinitions.keys()])
        // Set insertion order: usages order, then compose defs, then action defs.
        var usedOrder: [String] = []
        var usedSet = Set<String>()
        for name in index.usages.map(\.0) + index.composeDefinitions.map(\.0) + index.actionDefinitions.map(\.0)
        where !usedSet.contains(name) {
            usedSet.insert(name)
            usedOrder.append(name)
        }

        let undefinedNames = usedOrder.filter { !defined.contains($0) }
        let unusedNames = index.envDefinitions.map(\.0).filter { !usedSet.contains($0) }

        let usagesMap = index.usagesMap()

        var findings: [Finding] = []
        var seen = Set<String>()
        for undefinedName in undefinedNames {
            for unusedName in unusedNames {
                if !isLikelyTypo(undefinedName, unusedName) { continue }
                let pairKey = [undefinedName, unusedName].sorted().joined(separator: "\0")
                if seen.contains(pairKey) { continue }
                seen.insert(pairKey)

                let origins = usagesMap[undefinedName] ?? []
                findings.append(makeFinding(
                    "typo", .warning, undefinedName,
                    "did you mean \"\(unusedName)\"? (\(undefinedName) is referenced but not defined, \(unusedName) is defined but unused)",
                    Array(origins.prefix(3))
                ))
            }
        }
        return findings
    }
}
